--- xlhtml-0.3.9.6/xlhtml/xlhtml.c	Sun Nov 18 18:13:56 2001
+++ xlhtml/xlhtml/xlhtml.c	Sun Nov 25 15:31:01 2001
@@ -1315,6 +1315,7 @@
 ********************************************************************/
 static void main_line_processor(U16 opcode, U16 version, U32 count, U16 last, U8 data)
 {
+	static U16 last_string_formula_rcf[3];
 	U16 cont_opcode = 0;
 	
 	/* If first pass, reset stuff. */
@@ -1826,9 +1827,10 @@
 					}
 					else
 					{							/* String UNI */
-						strcpy((char *)calc_val, OutputXML ? "<NotImplemented/>String Formula"
-                                                       : "***String Formula" );
-						NotImplemented++;
+						last_string_formula_rcf[0] = r;
+						last_string_formula_rcf[1] = c;
+						last_string_formula_rcf[2] = f;
+						break;
 					}
 				}
 				else
@@ -1841,6 +1843,36 @@
 				add_wb_array(r, c, f, opcode, (U16)0, calc_val, (U16)strlen((char *)calc_val), 0, 0);
 			}
 			break;
+		case 0x07:	/* string formula result FIXME: test UNICODE and handle string formatting*/
+			working_buffer[bufidx++] = data;
+			if (bufidx == last)
+			{
+				U16 r = last_string_formula_rcf[0];
+				U16 c = last_string_formula_rcf[1];
+				U16 f = last_string_formula_rcf[2];
+				U8* str;
+				U16 length;
+				int unicode = 0;
+
+				length = working_buffer[0];
+
+				if (file_version == EXCEL97)
+				{
+					U16 try_len = (U16)working_buffer[0] | (U16)(working_buffer[1] << 8);
+					str = &working_buffer[3];
+
+					if (((try_len * 2) + 3) == last)	/* It's a UNICODE length */
+					{
+						length  = try_len;
+						unicode = 1;
+					}
+				}
+				else
+					str = &working_buffer[1];
+
+				add_wb_array(r, c, f, opcode, unicode, str, length, 0, 0);
+			}
+			break;
 		case 0x5C:	/* Author's name A.K.A. WRITEACCESS */
 			working_buffer[bufidx++] = data;
 			if ((bufidx == last)&&(author.str == 0))
@@ -3584,33 +3616,16 @@
 					printf("%g", dnum);
 					break;
 				default:	/* Unsupported...but, if we are here, its a number */
-					{
-						char *ptr = strchr((char *)u->str, '.');
-						if( OutputXML )
-							printf( "<NoFormat/>" );
-						if (ptr)
-						{
-							dnum = atof((char *)u->str);
-							if (Csv)
-								printf("\"%.15g *\"", dnum );
-							else if (OutputXML)
-								printf("%.15g", dnum );
-							else
-								printf("%.15g *", dnum );
-						}
-						else
-						{
-							if (Csv)
-								printf("\"%d *\"", atoi((char *)u->str) );
-							else if (OutputXML)
-								printf("%d", atoi((char *)u->str) );
-							else
-								printf("%d *", atoi((char *)u->str) );
-						}
-							
-/*						printf(" F:%02X", idx); */
-						NoFormat++ ;
-					}
+					if( OutputXML )
+					printf( "<NoFormat/>" );
+					dnum = atof((char *)u->str);
+					if (Csv)
+						printf("\"%.15g *\"", dnum );
+					else if (OutputXML)
+						printf("%.15g", dnum );
+					else
+						printf("%.15g *", dnum );
+					NoFormat++ ;
 					break;
 			}
 		}
